components/movementsensor/adxl345/adxl345.go

```diff
@@ -254,13 +254,14 @@ func NewAdxl345(
 					sensor.mu.Unlock()
 					continue
 				}
-
-				linearAcceleration := toLinearAcceleration(rawData)
-				// Only lock the mutex to write to the shared data, so other threads can read the
-				// data as often as they want.
-				sensor.mu.Lock()
-				sensor.linearAcceleration = linearAcceleration
-				sensor.mu.Unlock()
+				if len(rawData) >= 6 {
```

Comment:
This fixes the race condition, but is there a better check to be making here? Something that would signify that the rawData can be safely converted to linearAcceleration?

Reply:
it's the same thing we do in `movementsensor/imuwit/imuwit.go` and `movementsensor/gpsnmea/pmti2c.go`, checks are good to figure out that we're not getting corrupted data.

@penguinland what would be the best check be here to signal that we're getting good data off the movementsensor and also fulfill that race condition sending bad data over a test?

Also, since this is changing the driver, we need to test it out. Is there a way to mock this out with an injected i2c object in that test?

Reply:
Yes, I think we can. I will add one in!

Reply:
Thanks to Rand for pushing me to look into this further! I now suspect this if statement solves things because the mocked-out I2C handle doesn't act the way the real one should. The real one ought to return either the correct number of bytes of data or a non-nil error, but line 36 of adxl345_test.go will return 1 byte of data (and a nil error) no matter what.

The approach that most closely mirrors my personal model of how this works is to modify the `ReadBlockDataFunc` in the mock I2C handle so that it always returns the amount of data requested of it, and then we can remove this if statement. However, I recognize that I'm proposing a longer change than the one you've created.

Reply:
> since this is changing the driver, we need to test it out

If my other comments are right (and please double-check that, since I'm not positive!), the way to fix this is to modify the test and leave the driver alone. 

> Is there a way to mock this out with an injected i2c object in that test?

Ironically, the problem is due to mocking it out with an injected i2c object, and the real one shouldn't have these problems at all! (but again, please double-check my work. and if it doesn't make sense to anyone else, I could totally be wrong.)

Reply:
Oh, i see! This fixed it as well. Thanks Alan